test-runner-frontend/src/Views/DashboardUtils.js

```diff
@@ -0,0 +1,185 @@
+      mode: testMethod.mode,
+    };
+    return map;
+  }, {});
+};
+
+export const generateComparisonBetween = (currentTestRun, previousTestRun) => {
+  const currentTestRunModeAndScore = currentTestRun.statistics?.map((testMethod) => [
+    testMethod.benchmark,
+    testMethod.mode,
+    testMethod.primaryMetric.score,
+  ]);
+
+  const previousTestRunModeAndScore = convertToMap(previousTestRun);
+
+  let improvementsAndRegressions = currentTestRunModeAndScore
+    ?.map(([benchmark, mode, score]) => {
+      const comparisonBenchmark = previousTestRunModeAndScore[benchmark];
+      const comparisonMode = previousTestRunModeAndScore[benchmark]?.mode;
+      const comparisonScore = previousTestRunModeAndScore[benchmark]?.score;
+      if (comparisonBenchmark && comparisonMode === mode && score !== 0) {
+        return compare(score, comparisonScore, mode);
+      }
+      return null;
+    })
+    .filter((testRun) => testRun);
+
+  var counter = 0;
+
+  const improvedVsRegressed = {
+    improved: improvementsAndRegressions
+      ?.filter((improved) => improved === 'improved')
+      .reduce((first, second) => first + 1, counter),
+    regressed: improvementsAndRegressions
+      ?.filter((regressed) => regressed === 'regressed')
+      .reduce((first, second) => first + 1, counter),
+  };
+
+  return improvedVsRegressed;
+};
+
+export const getBenchmarksByMachine = (flattenedTestRuns) => {
+  const benchmarksByMachine = flattenedTestRuns.map((testRunsSortedByMachine) => {
+    return testRunsSortedByMachine
+      .map((testRun, index) => {
+        if (testRunsSortedByMachine[index + 1] && testRunsSortedByMachine[index + 1].statistics) {
+          const getImprovedVsRegressedValues = generateComparisonBetween(
+            testRun,
+            testRunsSortedByMachine[index + 1] ? testRunsSortedByMachine[index + 1] : {}
+          );
+          testRun.improvedVsRegressed.improved = getImprovedVsRegressedValues?.improved
+            ? getImprovedVsRegressedValues.improved
+            : 0;
+          testRun.improvedVsRegressed.regressed = getImprovedVsRegressedValues?.regressed
+            ? getImprovedVsRegressedValues.regressed
+            : 0;
+        }
+        return {
+          id: testRun.id,
+          commitHash: testRun.commitHash,
+          branch: testRun.branch,
+          status: testRun.status,
+          benchmarks: testRun.benchmarks,
+          machine: testRun.machine,
+          date: testRun.date,
+          statistics: testRun.statistics,
+          improvedVsRegressed: testRun.improvedVsRegressed,
+        };
+      })
+      .filter((testRun) => testRun);
+  });
+  return benchmarksByMachine;
+};
+
+export const getAllBenchmarks = (benchmarksByMachine) => {
+  const allBenchmarks = benchmarksByMachine.flatMap((benchmarkByMachine) => {
+    return benchmarkByMachine
+      .map((testRun) => {
+        return testRun;
+      })
+      .filter((testRun) => testRun);
```

Comment:
im not too sure i understand what is going on here. `benchmarksByMachine` is an array of an array of benchmarks (`Array<Array<Benchmarks>>`). we flat map the results of an array that are mapped to its current value and are not falsey.

i dont think we need to do the `.map()` and are the benchmarks ever not truthy? can we replace the the `.flatMap()` with just `.flat()`?

Reply:
Wow that turned the entire function into 1 line! And, no, I think at this point benchmarks can't be falsey so there's no need for the filter() method.